### DISCUSSION TITLE
Allow Shaders / Materials for Decals

### DISCUSSION THREAD

**Discussion by Travistyse** [godotengine/godot-proposals#4636] [3D]
Changing how the Decal blends with the environment in a custom way is incredibly helpful and I'll offer some examples as to where this would be useful:

Stylized lights and shadows that adhere to a specific color palette and falloff type (this can be integrated as additional parameters)

Procedural effects akin to particles (think swirling shadows, blood, water, dust) (I'm not sure how else this could be exposed to particles

Currently the only way to make this happen is using flipbooks and that's pretty limited.

I don't use ORM because I'm making a stylized game but I will note that even the ORM Material section just shows you options for Textures containing ORM information rather than ORM 3D Materials which can use custom shaders. 

Edit: Adjusting Decal Scale based on projection distance would be great for spot shadows

**Comment by aXu-AP** [comment]
I haven't yet used decals, but I imagine some cases can be handled by rendering effects into ViewportTexture.

From technical viewpoint, how would this shader work? Ie. what goes in, what goes out?
I imagine decal shader itself is already quite specialized in what it does, so abstracting some of it for custom shaders would be hard. On the other hand I don't know a lot about the subject 😄 

Cone shaped decals would definitely be cool also for projectors.

**Reply by Travistyse** [reply]
I think what I'm looking for is two separate potential features:

1. The ability to feed the Decal with a texture that I've created / modified via a shader. Certainly I'd prefer more directly being able to pass in a full material, but this would work fine nonetheless.

2. The ability to change the Mesh being projected - not very important but can help with some effects / performance / scaling issues (ie a large good looking circular edge needs a decently high resolution texture but the same can be managed with a circle mesh)

The main reason why I don't believe that I can just use a Viewport shader afterwards is mainly covering situations involving effects which need to take into account both the scene beneath the decal and the location of the decal's pixels. Mostly to do with blending but.. it kind of sounds like I might be better served by looking to integrate one of the 3.x Godot decal shaders in those cases.

**Comment by Calinou** [comment]
The decal implementation in 4.0.alpha is fast *because* it never performs material/shader changes. If this was changed to allow for custom materials/shaders, it would no longer be quite as fast.

If you need custom shaders, I suggest using PlaneMesh with a transparent material applied to it instead of using a Decal node. You can use a [reverse proximity fade trick](https://github.com/godotengine/godot-proposals/issues/3021) to fade the decal away when it gets far away from a solid surface.

> Edit: Adjusting Decal Scale based on projection distance would be great for spot shadows

This can be achieved using a RayCast node that adjusts the decal's X and Z extents based on the RayCast's hit position relative to the dynamic object's origin.

**Reply by Travistyse** [reply]
I think that the worst part about the suggestion to manipulate the extents is that I'm.. already doing that and forgot.

Maybe an "Advanced Decal" ? To me, a Decal only gets used when you need projection. I actually already have a spot shadow working using a circle mesh and a raycast but as the project continued to progress I decided to "finish" the feature by making it a decal which causes it to project off of cliffs and not just be a circle on the ground beneath grass and flowers and whatnot. 

It's technically feasible to split the shadow mesh into like.. 32 pieces or something nuts and raycast for each of them and deform it that way.. or use a bunch of bones lol. But if what I'm going for is posterchild decals, why go through all of that effort?

Nonetheless, it's completely fine for Godot's built in decals to trade utility for performance, but in my experience those "utility" features tend to be the eyecandy that sells a mechanic, thematic, or environment and so for me these simply aren't enough. But I'm completely satisfied with the shader decals in those cases so I've got a workaround (: 

But yeah, projected textures let you be far more creative with runtime content creation and can potentially save performance and dev time. It's a powerful feature and I'm glad that Godot features it even if I'm still reaching for the shader based decals half the time - especially given that my suggestion to expose different versions of Decals with differing features/parameters probably isn't that desirable for the engine team. A plane mesh just won't do what a projected texture can without a lot of effort.

**Reply by nonunknown** [reply]
I dissagre, an easy solution its create 2 kinds of decal, something like "static/dynamic" maybe? the static is the actual one, and the dynamic is the one that support shaders/materials
